dev_tools/check_shellscript_set_options.py

```diff
@@ -12,22 +12,46 @@
     from pathlib import Path
 
 
-def _sets_options_or_is_nolint(line: str) -> bool:
-    return line.strip() in ["set -euxo pipefail", "# nolint(set_options)"]
+def _sets_options_or_is_nolint(line: str, expected_options: str) -> bool:
+    return line.strip() in [ expected_options, "# nolint(set_options)"]
 
 
-def _is_valid_bash_file(filename: Path) -> bool:
-    return any(_sets_options_or_is_nolint(line) for line in filename.open().readlines())
+def _is_valid_shell_file(filename: Path, expected_options: str) -> bool:
+    return any(_sets_options_or_is_nolint(line, expected_options) for line in filename.open().readlines())
+
+
+def _separate_bash_from_sh_files(filenames: Sequence[Path]) -> tuple[list[Path], list[Path]]:
+    bash_files = []
+    sh_files = []
+    for filename in filenames:
+        first_line = filename.open().readline()
+        if first_line.startswith(("#!/bin/bash", "#!/usr/bin/bash")):
+            bash_files.append(filename)
+        elif first_line.startswith(("#!/bin/sh", "#!/usr/bin/sh")):
+            sh_files.append(filename)
+        else:
+            msg = f"Unknown shell in {filename}: {first_line}. Only use this hook in combination with 'check-executables-have-shebangs' from https://github.com/pre-commit/pre-commit-hooks"
+            raise ValueError(msg)
+
+    return bash_files, sh_files
+
+
+def _are_shell_files_valid(shell_files: list[Path], expected_options :str) -> None:
+    invalid_shell_files = [filename for filename in shell_files if not _is_valid_shell_file(filename, expected_options)]
```

Comment:
Here a simple for loop that prints and sets `failure = True` or similar would be more readable.

Reply:
I'm fine either way - I'll let you fight over that with our main list comprehender @hofbi ;)

Reply:
All in for list comprehension

Reply:
It is not just about the filter loop. It is about the whole function.

```python
def _are_shell_files_valid(shell_files: list[Path], expected_options :str) -> None:
    valid = True
    filename for filename in shell_files:
        if not _is_valid_shell_file(filename, expected_options):
            print(f"Error: {filename} does not contain '{expected_options}'")
            valid = False
    return valid
```

Reply:
With your proposal you are mixing 2 things, filtering and printing. The list comprehension seperates concerns. You first filter and then print.